### PR TITLE
fix(proxy): 将 messages 中的 role:"system" 消息合并到顶层 system 字段，修复严格上游 400

### LINKED ISSUE #1265: Claude Code 2.1.172+ 自定义模型名时在 messages 中注入 role:"system" 消息，严格上游(Vertex/Bedrock类)拒绝导致重试熔断 503
## 问题描述

Claude Code **2.1.172** 配合**自定义模型名**（通过 `ANTHROPIC_DEFAULT_OPUS_MODEL` 等指向 cchub 渠道模型，如 `aws-claude-opus`）使用时，请求经 cchub 转发到严格校验的上游（Vertex / Bedrock 类 Anthropic 兼容端点）会被拒绝。cchub 重试耗尽全部供应商后触发熔断，客户端最终收到：

```json
{"statusCode": 503, "errorMessage": "所有供应商暂时不可用，请稍后重试"}
```

首次请求的流式响应为：

```
event: error
data: {"type": "error", "error": {"type": "api_error", "message": "Upstream returned an error event (request_id: ...)"}}
```

由于错误被包装成 `api_error` / 503，且重试导致熔断殃及后续请求，排查非常困难。

## 根因（已通过本地抓包确认）

Claude Code 2.1.172（2.1.153 之后引入的行为变化）会根据**模型名是否为内置官方 Claude 模型 ID** 切换 skills 列表的注入格式：

| 模型名 | skills 注入方式 |
|---|---|
| 官方 ID（`claude-opus-4-7` 等） | user 消息内 `<system-reminder>` 文本块（旧行为） |
| 任何非官方名（`aws-claude-opus`、甚至 `claude-test-fake-9`） | **`messages` 数组中独立的 `{"role": "system", "content": "..."}` 消息** |

抓包验证（同一 CC 版本、本地 HTTP 捕获服务器）：

```
--model claude-opus-4-7   → messages roles: ['user']            ✅
--model aws-claude-opus   → messages roles: ['user', 'system']  ❌
--model claude-test-fake-9 → messages roles: ['user', 'system'] ❌
```

Anthropic 官方 API 接受 messages 中的 `role: "system"`，但 Vertex / Bedrock 等转换层严格校验 messages 角色只能为 `user` / `assistant`，直接返回 400 类错误（"Bad request - check request format and parameters"）。cchub 目前原样透传该消息，导致：

```
CC 2.1.172（自定义模型名）发出 role:"system" 消息
    → cchub 原样转发
    → 严格上游 400 拒绝
    → cchub 重试其余供应商全部失败
    → 熔断 → 503 "所有供应商暂时不可用"
```

由于同一模型可能挂多个渠道（宽容的渠道能成功），故障表现为**间歇性**，进一步增加排查难度。

## 影响范围

所有满足以下条件的用户：

- Claude Code ≥ 2.1.172（2.1.153 确认无此行为，具体引入版本未逐版定位）
- 渠道模型名为自定义命名（如 `aws-claude-opus`、`qwen-xx`、`glm-xx`、`MiniMax-xx` 等，这是 cchub 用户的常见用法）
- 上游为严格校验 messages 角色的 Anthropic 兼容端点

## 复现步骤

```bash
# 对任意 Vertex/Bedrock 类严格上游渠道发送：
curl -s "https://<cchub>/v1/messages" \
  -H "x-api-key: $KEY" -H "anthropic-version: 2023-06-01" \
  -H "content-type: application/json" \
  -d '{
    "model": "<自定义模型名>", "max_tokens": 100,
    "messages": [
      {"role": "user", "content": "hi"},
      {"role": "system", "content": "The following skills are available: ..."}
    ]
  }'
# → 400/503；去掉 role:"system" 那条消息后同样请求成功
```

或直接用 Claude Code 2.1.172 + `ANTHROPIC_MODEL=<自定义模型名>` 发送任意消息（只要本地启用了任意 skill/plugin 就会注入该消息）。

## 建议修复

参照现有 `src/app/v1/_lib/proxy/billing-header-rectifier.ts` 的模式（同样是处理 Claude Code 新版本注入的、上游不兼容的内容），新增一个 proactive rectifier：

- 在转发前扫描 `messages`，将 `role: "system"` 的消息取出
- 其内容合并到顶层 `system` 数组末尾（string content 转为 `{type: "text", text}` block）
- 语义上与 Anthropic 官方对该写法的处理一致，对宽容上游无副作用，对严格上游消除 400

如方案可接受，我可以提交 PR。

## 环境

- claude-code-hub: v0.8.5（self-hosted）
- Claude Code: 2.1.172（user-agent: `claude-cli/2.1.172 (external, cli)`）
- 对照版本: Claude Code 2.1.153 无此问题

### PR DESCRIPTION
**Pull request by 00010110**
## 背景

Fixes #1265

Claude Code 2.1.172+（2.1.153 之后引入的行为变化）配合自定义模型名（`ANTHROPIC_DEFAULT_OPUS_MODEL` 等指向 cchub 渠道模型）使用时，会在 `messages` 数组中注入独立的 `{"role": "system", "content": "..."}` 消息（skills 列表）。Anthropic 官方 API 接受该写法，但 Vertex/Bedrock 类严格校验 messages 角色的兼容上游会以 400 拒绝，cchub 原样透传导致重试耗尽全部供应商后熔断，客户端收到 503「所有供应商暂时不可用」。

## 修复方案

参照现有 `billing-header-rectifier.ts` 的 proactive rectifier 模式：

### 核心模块 `src/app/v1/_lib/proxy/system-message-rectifier.ts`

- 转发前扫描 `messages`，取出 `role: "system"` 条目
- string content 转为 `{type: "text", text}` block；数组 content 保留 text block（含 `cache_control` 等附加字段），丢弃非 text block
- 追加到顶层 `system` 字段末尾（顶层为 string 时先转为首个 text block）
- 原地变异，语义与 Anthropic 官方 API 对该写法的处理一致：宽容上游无行为变化，严格上游消除 400

### forwarder 集成

- 仅对 `claude` / `claude-auth` 供应商生效
- 置于 billing header rectifier **之前**，确保 system 消息内若携带 `x-anthropic-billing-header` 文本块，搬移后仍会被后续剥离
- 命中时记录 `system_message_rectifier` specialSetting 审计并写日志

### 系统设置开关 `enableSystemMessageRectifier`（默认开启）

- 全链路：schema 新列 + 迁移（`drizzle/0105_easy_cyclops.sql`）、settings cache、types、zod 校验、Server Action、admin REST、v1 OpenAPI schema、repository、transformers
- 按既有模式接入 `system_settings` 缺列降级链（读取与更新路径均在链头新增一层剔除该列的降级），迁移前部署的实例读取/更新设置不受影响
- 设置页新增开关项，补全 zh-CN / zh-TW / en / ja / ru 五语言文案

## 测试

新增 `tests/unit/proxy/system-message-rectifier.test.ts`，16 个用例：

- string / array content、空 content、非法 content
- 顶层 system 为 undefined / string / array / 空串 / 未知类型
- 多条 system 消息保序、messages 含 null / 非对象条目
- 原地变异引用不变
- 与 billing rectifier 按 forwarder 实际顺序的组合场景（system 消息内携带 billing header 块最终被剥离）

## 验证

- `bun run build` 通过
- `bun run lint` / `lint:fix` 通过
- `bun run typecheck` 通过
- `bun run test` 全量 6370 个测试通过
- `bun run openapi:check` 通过

<details open><summary><h3>Greptile Summary</h3></summary>

This PR adds a proactive rectifier that lifts `role:"system"` messages out of the `messages` array and merges them into the top-level `system` field before forwarding, matching the lenient interpretation the official Anthropic API already applies and removing the source of 400 errors on strict upstreams like Vertex/Bedrock.

- **Core rectifier** (`system-message-rectifier.ts`): synchronous, in-place mutation; correctly handles all content shapes (string → single block, array → filter to text blocks, empty string → no block, unknown type → discard), and all top-level `system` shapes (undefined, null, string, array, unknown).
- **Forwarder integration**: runs before the billing-header rectifier (so billing blocks moved via this pass are still stripped), guarded by `enableSystemMessageRectifier` setting; both rectifiers share the single `getCachedSystemSettings()` result fetched once per `doForward` call.
- **Settings plumbing**: full migration, cache, schema, zod, admin API, and admin UI coverage; column-missing fallback chain prepends the new column to the existing degradation ladder.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge. The rectifier is synchronous, idempotent, and only mutates messages when role:"system" entries are present — requests without them pass through unchanged.

The change is a well-bounded proactive transformation applied only to claude/claude-auth providers, guarded by a feature flag that defaults to the safe/on state. The core function handles all edge cases and is tested with 16 targeted unit tests. Settings plumbing follows the established pattern with the new column correctly positioned as the outermost entry in the degradation ladder.

No files require special attention.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/app/v1/_lib/proxy/system-message-rectifier.ts | New core rectifier module — well-typed, mutates in place, handles all content/system field shapes correctly; comprehensive unit tests. |
| src/app/v1/_lib/proxy/forwarder.ts | Integrates the new rectifier before the billing-header rectifier, both sharing a single getCachedSystemSettings() result. Logic and ordering are correct. |
| src/repository/system-config.ts | Adds enableSystemMessageRectifier to both full-select and fallback ladder (as the outermost/newest fallback step), and to the update path fallback; consistent with prior pattern. |
| drizzle/0106_stiff_dormammu.sql | Single-column migration adding enable_system_message_rectifier boolean NOT NULL DEFAULT true — safe, additive, non-blocking. |
| tests/unit/proxy/system-message-rectifier.test.ts | 16 well-chosen unit tests covering all content shapes, system field shapes, ordering, in-place mutation, and the combined billing-header interaction. |
| src/types/special-settings.ts | Adds SystemMessageRectifierSpecialSetting type; consistent shape with BillingHeaderRectifierSpecialSetting. |
| src/lib/config/system-settings-cache.ts | Adds enableSystemMessageRectifier: true to the DEFAULT_SETTINGS fallback and to the cold-cache fallback object; consistent with other rectifier defaults. |
| tests/unit/repository/system-config-update-missing-columns.test.ts | Existing test updated with the new fallback path; the test mock for the full-column set does not include the new enableSystemMessageRectifier field (minor gap, but not blocking). |

</details>

</details>

<details open><summary><h3>Sequence Diagram</h3></summary>

```mermaid
sequenceDiagram
    participant CC as Claude Code Client
    participant FW as ProxyForwarder.doForward
    participant SMR as SystemMessageRectifier
    participant BHR as BillingHeaderRectifier
    participant UP as Upstream (Vertex/Bedrock)

    CC->>FW: POST /v1/messages
    FW->>FW: getCachedSystemSettings()
    alt "enableSystemMessageRectifier = true"
        FW->>SMR: rectifySystemMessages(message)
        SMR-->>FW: "applied=true, messages cleaned, system merged"
    end
    alt "enableBillingHeaderRectifier = true"
        FW->>BHR: rectifyBillingHeader(message)
        BHR-->>FW: billing header blocks stripped
    end
    FW->>UP: POST with clean messages array
    UP-->>FW: 200 OK
    FW-->>CC: 200 OK
```
</details>

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `src/app/v1/_lib/proxy/forwarder.ts`, line 2480-2483 ([link](https://github.com/ding113/claude-code-hub/blob/b61e1d7855fb9de2949f8ae16f98c31eb7db5e1a/src/app/v1/_lib/proxy/forwarder.ts#L2480-L2483)) 

   <a href="#">P2</a> Two consecutive `getCachedSystemSettings()` calls are made in the same code path — one for the system message rectifier and one for the billing header rectifier. Since the cache returns the same in-memory object within the 1-minute TTL, the second call is a no-op but still adds an unnecessary `await`. Both rectifiers could share the single `settings` object that was already fetched.

   

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: src/app/v1/_lib/proxy/forwarder.ts
   Line: 2480-2483

   Comment:
   Two consecutive `getCachedSystemSettings()` calls are made in the same code path — one for the system message rectifier and one for the billing header rectifier. Since the cache returns the same in-memory object within the 1-minute TTL, the second call is a no-op but still adds an unnecessary `await`. Both rectifiers could share the single `settings` object that was already fetched.

   

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

   Note: If this suggestion doesn't match your team's coding style, reply to this and let me know. I'll remember it for next time!
</details>

<sub>Reviews (3): Last reviewed commit: ["merge: resolve conflicts with dev (PR #1..."](https://github.com/ding113/claude-code-hub/commit/b7619abb59ded1e083450d83d2152a9a945e5260) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=36902228)</sub>

### CODE REVIEW

File: src/app/v1/_lib/proxy/system-message-rectifier.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  applied: boolean;
+  movedCount: number;
+  extractedValues: string[];
+};
+
+const NOOP_RESULT: SystemMessageRectifierResult = {
+  applied: false,
+  movedCount: 0,
+  extractedValues: [],
+};
```
[the remark this answers is not in the document]
Comment (00010110) [reply src/app/v1/_lib/proxy/system-message-rectifier.ts]: 已在 1d2bd1a8 修复：`NOOP_RESULT` 常量已改为 `createNoopResult()` 工厂函数，每次调用返回独立的 `extractedValues` 数组，消除引用共享。

File: src/app/v1/_lib/proxy/forwarder.ts [hunk 2]
```diff
@@ -92,6 +92,7 @@ import { ProxyProviderResolver } from "./provider-selector";
 import { finalizeHedgeLoserBilling } from "./response-handler";
 import type { ProxySession } from "./session";
 import { setDeferredStreamingFinalization } from "./stream-finalization";
+import { rectifySystemMessages } from "./system-message-rectifier";
```
[the remark this answers is not in the document]
Comment (00010110) [reply src/app/v1/_lib/proxy/forwarder.ts:95]: 不采纳此建议：`forwarder.ts` 对同目录（`./proxy/`）模块一律使用相对导入，与既有的 `./billing-header-rectifier`、`./client-abort-listener` 等导入保持一致。仓库的 `@/` 别名约定用于跨目录导入，同目录文件改用 `@/` 反而会与本文件既有风格不一致。

**Comment by 00010110** [comment]
已根据审查意见更新（1d2bd1a8）：

- **CodeRabbit – NOOP_RESULT 共享数组引用**：已修复，改为 `createNoopResult()` 工厂函数，每次返回独立的 `extractedValues` 数组。
- **Greptile – 重复的 `getCachedSystemSettings()` 调用**：已修复，两个 rectifier 块现共用同一次 settings 读取。
- **CodeRabbit – `@/` 别名导入**：不采纳，本文件对同目录模块统一使用相对导入（与 `./billing-header-rectifier` 等保持一致），详见行内回复。

验证：`bun run build` / `typecheck` / 改动文件 lint 干净 / 全量 6370 测试通过。

**Comment by ding113** [comment]
hi，对于涉及请求体变更的整流器，建议按照思考签名整流器的思路，在检测到特定报错模式之后被动整流。
主动对请求体进行修改可能引发意料外后果。

**Comment by 00010110** [comment]
放弃此方向，关闭 PR。
